src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts

```diff
@@ -232,6 +234,19 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
     };
   }
 
+  fetchErasureCodeProfile(erasureCodeProfile: string) {
+    this.ecpService.list().subscribe((ecProfileList: ErasureCodeProfile[]) => {
+      _.forEach(ecProfileList, (ecpKey) => {
+        if (ecpKey['name'] === erasureCodeProfile) {
+          this.k = ecpKey['k'];
+          this.m = ecpKey['m'];
+        }
+      });
+    });
+
+    return `EC: ${this.k}+${this.m}`;
```

Comment:
It's luck that this works somehow. You're formatting a string with values that are not assigned at the current state because `this.ecpService.list().subscribe` is executed AFTER this code line. You're mixing synchronous and asynchronous code here.

Reply:
Yes  I see the problem. When the first it subscribes it shows 'undefined+undefined'. So how can I avoid this? Any suggestions?

Reply:
Will this work.. I just not return the string and use k & m values where the function is called (in transformsPoolData())?

Reply:
You better use a pipe for the `Data protection` column. This will make the asynchronicity of showing this data much more easier.

Reply:
async pipe?

Reply:
I tried with aysnc pipe but the issue is how to use it in template for `Data Protection` column? I looked if somehow async pipe can be used as column property but there doesn't seems to be one. So for now, I have removed the return statement and instead just used directly the updated values of k and m.

Reply:
@avanthakkar @votdev @epuertat IMHO we can solve 2 problems here:
1) No need to call this.ecpService.list() several times inside a loop: 1 is enough.
2) Async code issue can be resolved using ```mergeMap``` operator to ensure that we already have ecProfileList info when processing pool data.

The refactor can be something like this:
```
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts b/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
index 171d6871cce..f1c6c2c3d48 100644
--- a/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.ts
@@ -1,7 +1,7 @@
 import { Component, OnInit, TemplateRef, ViewChild } from '@angular/core';
 
 import _ from 'lodash';
-import { Observable } from 'rxjs';
+import { mergeMap } from 'rxjs/operators';
 
 import { PgCategoryService } from '~/app/ceph/shared/pg-category.service';
 import { ConfigurationService } from '~/app/shared/api/configuration.service';
@@ -60,9 +60,7 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
   tableStatus = new TableStatusViewCache();
   cacheTiers: any[] = [];
   monAllowPoolDelete = false;
-  k: number;
-  m: number;
-  ecProfileList$: Observable<ErasureCodeProfile[]>;
+  ecProfileList: ErasureCodeProfile[];
 
   constructor(
     private poolService: PoolService,
@@ -196,7 +194,13 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
     ];
 
     this.taskListService.init(
-      () => this.poolService.getList(),
+      () =>
+        this.ecpService.list().pipe(
+          mergeMap((ecProfileList: ErasureCodeProfile[]) => {
+            this.ecProfileList = ecProfileList;
+            return this.poolService.getList();
+          })
+        ),
       undefined,
       (pools) => {
         this.pools = this.transformPoolsData(pools);
@@ -236,15 +240,15 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
     };
   }
 
-  fetchErasureCodeProfile(erasureCodeProfile: string) {
-    this.ecpService.list().subscribe((ecProfileList: ErasureCodeProfile[]) => {
-      _.forEach(ecProfileList, (ecpKey) => {
-        if (ecpKey['name'] === erasureCodeProfile) {
-          this.k = ecpKey['k'];
-          this.m = ecpKey['m'];
-        }
-      });
+  getErasureCodeProfile(erasureCodeProfile: string) {
+    let ecpInfo = '';
+    _.forEach(this.ecProfileList, (ecpKey) => {
+      if (ecpKey['name'] === erasureCodeProfile) {
+        ecpInfo = `EC: ${ecpKey['k']}+${ecpKey['m']}`;
+      }
     });
+
+    return ecpInfo;
   }
 
   transformPoolsData(pools: any) {
@@ -283,8 +287,7 @@ export class PoolListComponent extends ListWithDetails implements OnInit {
 
       if (pool['type'] === 'erasure') {
         const erasureCodeProfile = pool['erasure_code_profile'];
-        this.fetchErasureCodeProfile(erasureCodeProfile);
-        pool['data_protection'] = `EC: ${this.k}+${this.m}`;
+        pool['data_protection'] = this.getErasureCodeProfile(erasureCodeProfile);
       }
       if (pool['type'] === 'replicated') {
         pool['data_protection'] = `replica: ×${pool['size']}`;
```

Reply:
Looks good to me..tested it and works perfect! Thanks @alfonsomthd 
Just one thing can't we use switchMap too here? As all we need is just the latest value, but I think it doesn't make diference as we are gonna do just 1 call , right?

Reply:
@avanthakkar for this particular case it does not make a difference as httpClient is just returning 1 value.
```switchMap``` is usually used when you also need cancellation logic, which is not the case.